analyzers/src/SonarAnalyzer.Common/Rules/UseIndexingInsteadOfLinqMethodsBase.cs

```diff
@@ -0,0 +1,74 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Rules;
+
+public abstract class UseIndexingInsteadOfLinqMethodsBase<TSyntaxKind, TInvocation> : SonarDiagnosticAnalyzer<TSyntaxKind>
+    where TSyntaxKind : struct
+    where TInvocation : SyntaxNode
+{
+    private const string DiagnosticId = "S6608";
+    protected override string MessageFormat => "{0} should be used instead of the \"Enumerable\" extension method \"{1}\"";
+
+    private static readonly ImmutableArray<KnownType> TargetInterfaces = ImmutableArray.Create(
+        KnownType.System_Collections_IList,
+        KnownType.System_Collections_Generic_IList_T);
+
+    protected abstract int GetArgumentCount(TInvocation invocation);
+    protected abstract bool TryGetOperands(TInvocation invocation, out SyntaxNode left, out SyntaxNode right);
+    protected abstract SyntaxToken? GetIdentifier(TInvocation invocation);
+
+    protected UseIndexingInsteadOfLinqMethodsBase() : base(DiagnosticId) { }
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterNodeAction(Language.GeneratedCodeRecognizer, c =>
+        {
+            CheckInvocation(c, nameof(Enumerable.First), 0, "0");
```

Comment:
I don't think I really like this approach.
Why? If `First()` is found you still check whether the invocation is `Last` or `ElementAt`, which is a bit unnecessary (it can be only one). Same for the node casting, which is done 3 times for every invocation.